xdsl/passes.py

```diff
@@ -1,20 +1,123 @@
 from abc import ABC, abstractmethod
+from dataclasses import dataclass, Field
+from types import UnionType, NoneType
+
+from xdsl.utils.hints import isa
 from xdsl.dialects import builtin
 from xdsl.ir import MLContext
-from typing import ClassVar
+from typing import ClassVar, TypeVar, get_origin, Any, Union, get_args
+from xdsl.utils.parse_pipeline import (
+    PipelinePassSpec,
+    PassArgListType,
+    PassArgElementType,
+)
+
+_T = TypeVar("_T", bound="ModulePass")
 
 
+@dataclass
 class ModulePass(ABC):
     """
     A Pass is a named rewrite pass over an IR module.
 
     All passes are expected to leave the IR in a valid state after application.
     That is, the IR verifies. In turn, all passes can expect the IR they are
-    applied to to be in a valid state.
+    applied to be in a valid state.
     """
 
     name: ClassVar[str]
 
     @abstractmethod
     def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
         ...
+
+    @classmethod
+    def from_pass_spec(cls: type[_T], spec: PipelinePassSpec) -> _T:
+        """
+        This method takes a PipelinePassSpec, does type checking on the
+        arguments, and then instantiates an instance of the ModulePass
+        from the spec.
+        """
+        # some sanity checks here
+        assert issubclass(cls, ModulePass), f"{cls} must be subclass of ModulePass"
+        assert hasattr(cls, "__dataclass_fields__"), f"{cls} must be a dataclass"
+        assert spec.name == cls.name, "Wrong pass name provided"
+
+        # normalize spec arg names:
+        spec.normalize_arg_names()
+
+        # get all dataclass fields
+        fields: dict[str, Field[_T]] = cls.__dataclass_fields__
+
+        # start constructing the argument dict for the dataclass
+        arg_dict = dict[str, PassArgListType | PassArgElementType | None]()
+
+        # iterate over all fields of the dataclass
+        for field in fields.values():
+            # ignore the name field and everything that's not used by __init__
+            if field.name == "name" or not field.init:
+                continue
+            # check that non-optional fields are present
+            if field.name not in spec.args and not _is_optional(field.type):
+                raise Exception(f'Pass {cls.name} requires argument "{field.name}"')
+
+            # convert pass arg to the correct type:
+            arg_dict[field.name] = _convert_pass_arg_to_type(
+                # we use default value [] here, this makes handling optionals easier
+                spec.args.pop(field.name, []),
+                field.type,
+            )
+            # we use .pop here to also remove the arg from the dict
+
+        # if not all args were removed we raise an error
+        if len(spec.args) != 0:
+            raise Exception(f'Unrecognised pass argument "{list(spec.args)[0]}"')
```

Comment:
```suggestion
            raise Exception(f'Unrecognised pass arguments "{list(spec.args)}"')
```
Maybe worth printing them all instead of one by one?

Reply:
You're right. I hate that when a program does that to me :)